update docs with coding standards (#10214)
bbc/simorgh: docs/Coding-Standards/React-Components.md

  margin: 1rem; // spacing around component
  width: 300px; // restricted width
  background-color: #f5f5f5;
`;

const Main = () => (
  <main>
    <StyledPromosContainer>
      <StyledPromo style={{ width: '100%', margin: 0 }} />
      <StyledPromo style={{ width: '100%', margin: 0 }} />
      <StyledPromo style={{ width: '100%', margin: 0 }} />
    </StyledPromosContainer>
  </main>
);
```

✅

```jsx
const StyledPromosContainer = styled.div`
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 1rem;
`;

const StyledPromo = styled.div`
  background-color: #f5f5f5;
`;

const Main = () => (
  <main>
    <StyledPromosContainer>
      <StyledPromo />
      <StyledPromo />
      <StyledPromo />
    </StyledPromosContainer>
  </main>
);
```

## Keep styled components in a separate file

### Why?

Component files are easier to reason about when you separate the presentation from the content and behaviour aspects of a component. With this separation of concerns, there is less cognitive load when it comes to updating the content or behaviour of a component because everything concerning the presentation is out of sight and mind in a separate file.

### How?

> Put styled components in a file with the same filename as the component but with `.styled` before the `.jsx` file extension. For example, `index.styled.jsx`. Then reference the styled components in the component file.

Reviewer: ```suggestion
Put styled components in a file with the same filename as the component but with `.styled` before the `.tsx` file extension. For example, `index.styled.jsx`. Then reference the styled components in the component file.
```

> Put styled components in a file with the same filename as the component but with `.styled` before the `.jsx` file extension. For example, `index.styled.jsx`. Then reference the styled components in the component file.

Reviewer: ```suggestion
Put styled components in a file with the same filename as the component but with `.styled` before the `.tsx` file extension. For example, `index.styled.jsx`. Then reference the styled components in the component file.
```

❌

```jsx
const StyledWrapper = styled.div`

>   display: grid;

Reviewer 2: might want to change this example from a grid layout as it contradicts our other standard around the use of the grid layout for anything other than page layout

Author: This example is a page layout. I can still change if it if you like though.

const Page = () => (
  <StyledWrapper>
    <StyledMain />
    <StyledSupplementaryContent />
  </StyledWrapper>
)
```

## Primitive Components, Compound Components and Complex Components

For our frontend components, we focus on reuse at root of our React Component tree. We have a set of ‘Primitive Components’ that are the building blocks of other components higher up the tree.

### Primitive Components

These components will have few if any internal dependencies and focus on applying common concerns universal everywhere such as typography.

In a high level example below we define some building blocks of a promo including `Title` , `Timestamp` and `Image` .

### Compound Components

These components apply the ‘[Compound Component Pattern](https://kentcdodds.com/blog/compound-components-with-react-hooks)’ to create a ‘Compound Component’ that can be used to implement a particular user experience across different page types.

In our worked example we create a `Promo` that allows customised versions of `Title` , `Timestamp` and `Image` to be used in different ways depending on the use case.

### Complex Components

> These components are components that use Primitive components to build a reusable component that isn’t suitable to be implemented as a Compound Component. It needs limited or no customisation for usages on different pages. A Complex Component could later be converted to Compound Component if it needed significant customisation that differed across Page Types.

Reviewer 3: ```suggestion
These components use Primitive components to build a reusable component that isn’t suitable to be implemented as a Compound Component. It needs limited or no customisation for usages on different pages. A Complex Component could later be converted to Compound Component if it needed significant customisation that differed across Page Types.
```

In our example below we have a `Byline` component used on both Article and Story Pages with no customisation at all.

### Applied Compound Components

> At this level we take a ‘Compound Component’ and build an ‘Applied Compound Component’ that is very specific to the page’s use case. At this level we can apply specific styling and add in additional functionality completly bespoke to that page’s requirements; we are:

Reviewer 3: ```suggestion
At this level we take a ‘Compound Component’ and build an ‘Applied Compound Component’ that is very specific to the page’s use case. At this level we can apply specific styling and add in additional functionality completely bespoke to that page’s requirements; we are:
```

- Reusing the functionality common across all pages from the Primitive Components
- Reusing the functionality common to all usages of a particular user experience from the Compound Component
- Applying customisations for the specific usage at the page level; we are free to make changes here without worrying about reuse

In our example these are `RelatedContentPromo` and `TopStoriesPromo` respectively. Differences to the Promos on the respective page types are applied in these Components and they are co-located with the Page.

### Bringing it Together

Below is the full worked example. The example shows our Article and Story pages implementing their own Applied Components for Related Content and Top Stories. They also use a Complex Component to implement a Byline at the top of the article, it needs no customisation between the page types. The Promo Compound Component and Byline Complex Component use the Primitive Components to present a resuable component for use across page types.

> ![](https://paper-attachments.dropbox.com/s_2A43B8E3F9C1F972929EDC9878EAFF3FE87CC9300DB1C147F888F337C763ADF5_1656492576866_file.png)

Reviewer 2: can these images be added to source control as these attachement links may not work long term? Also I can provide the draw.io original to source control with this doc:
[compound-components.drawio.zip](https://github.com/bbc/simorgh/files/9243214/compound-components.drawio.zip)

### Summary

> In the past we built, highly configurable, reusable components that required large amounts of props to be ‘drilled’ into component to allow them to work on a particular page; a particular example of this was the [StoryPromo](https://github.com/bbc/simorgh/blob/d025086b41fb8753ec1a3992977bef5b013410b7/src/app/containers/StoryPromo/index.jsx#L126) component. With these highly configurable components the complexity grew and grew as they were reused in more places.

Reviewer 3: ```suggestion
In the past we built highly configurable, reusable components that required large amounts of props to be ‘drilled’ into components to allow them to work on a particular page. An example of this was the [StoryPromo](https://github.com/bbc/simorgh/blob/d025086b41fb8753ec1a3992977bef5b013410b7/src/app/containers/StoryPromo/index.jsx#L126) component. With these highly configurable components the complexity grew and grew as they were reused in more places.
```

In the new approach we make simple highly reusable primitive components with few customisation options, we create CompoundComponents to guide reuse through a set of child components and then we apply these for our specific page types, pushing specialisation to the top of the component hierarchy.

## Component Locations

### Legacy Components

> Before looking to apply our component standards we have a huge body of components that will not meet the existing standards to provide a ‘clean slate’ these can be moved to a legacy location in the following manner:

Reviewer 2: Should this be updated to be in the present tense now but record for posterity where they all came from?

N.B. As referred to above we are moving away from the [c](http://)[omponent-container](https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0) pattern as engineers historically struggled to differentiate between the two and the distinction offered little from a maintenance perspective

### Standardised Components

With a clear understanding of where our ‘legacy’ components can be found we now make space for our ‘standardised’ components that meet our new standards agreed in this doc.

Firstly, we will move pages into the now empty components directory:
`/src/app/pages` → `/src/app/components/pages`

And all components meeting our standards will from now on will be maintained in the following directories:

| Component Types                 | Location                    |
| ------------------------------- | --------------------------- |
| Primitive, Compound, Complex    | `src/app/components`        |
| Applied Compound, Page Specific | `/src/app/components/pages` |

### Applying Standardisation

> Where possible we want apply standardisation as an ‘opportunitistic refactor’. As we look to use existing components to deliver a feature, we commit to rebuilding those components or porting them across applying new best practices. Through doing this we gradually improve the consistency and standards of our components rather then embarking on a unrealistic drive to rebuild all components to a new standard. The necessity of delivering a feature keeps scope under control while the standards maintain the quality of the code for that feature.

Reviewer 3: ```suggestion
Where possible we want to apply standardisation as an ‘opportunistic refactor’. As we look to use existing components to deliver a feature, we commit to rebuilding those components or porting them across applying new best practices. Through doing this we gradually improve the consistency and standards of our components rather then embarking on a unrealistic drive to rebuild all components to a new standard. The necessity of delivering a feature keeps scope under control while the standards maintain the quality of the code for that feature.
```

## Typescript for Frontend Components

> - Our visual components should be written in typescript clearly document each component’s API and get benefits of compile-time type checking between components.

Reviewer 3: ```suggestion
- Our visual components should be written in TypeScript to clearly document each component’s API and get the benefits of compile-time type checking between components.
```

- Types should be co-located with the component that uses them; they should only be maintained in a shared location where they are used by multiple visual components
  - Here is an example of shared types in our BFF application: https://github.com/bbc/fabl-modules/tree/master/modules/application/simorgh-bff/src/models
  - We can locate any shared types in `src/app/components/models`
- Types should be as strict as possible
  - Avoid use of the `any` type and try and define a more specific type

>   - Where a prop has a fixed set of options use a `[union](https://camchenry.com/blog/typescript-union-type)` [type](https://camchenry.com/blog/typescript-union-type) [](https://camchenry.com/blog/typescript-union-type)to restrict the prop to those fixed options

Reviewer 2: We should comment explicity on the inability to use aliases here I think

## Component Health

All components should have a component health visualisation in Storybook similar to [webcore](https://paper.dropbox.com/doc/What-is-component-health--BfG_Ski_wr7eLwwMF_aQZlIAAg-2qI1jQCZhKoaGVX4TVlCf?noDesktopRedirect=1)’s implementation:

![](https://paper-attachments.dropbox.com/s_0A228B6278B754FD729F6444013C8CCE0645C344FE8C5686DEE4A9889366E8C3_1623683012926_Screenshot+2021-06-14+at+16.03.01.png)

Component Health should include:

- Link to screen reader UX
- Accessibility acceptance criteria
- Accessibility swarm document
- Whether the component has ‘right-to-left’ support

> All docs linked too must be secure and not publically editable

Reviewer 3: ```suggestion
All linked docs must be secure and not publicly editable
```

## Readme

All components should have a readme that includes

- A plain text description of what the component does
- API docs describing the component’s props
- A11y notes: explaination of implmentation details specific to a11y requirements
- Usage examples including RTL
- List any third party dependencies in use

> - An explaination of any unsual/uninuitive implementation decisions required to work around an external bug (e.g. browser bug) or to deliver quickly to a deadline

Reviewer 2: ```suggestion
- An explaination of any unsual/unintuitive implementation decisions required to work around an external bug (e.g. browser bug) or to deliver quickly to a deadline
```

> - An explaination of any unsual/uninuitive implementation decisions required to work around an external bug (e.g. browser bug) or to deliver quickly to a deadline

Reviewer 3: ```suggestion
- An explanation of any unusual/uninuitive implementation decisions required to work around an external bug (e.g. browser bug) or to deliver quickly to a deadline
```
